labgrid/util/ssh.py

```diff
@@ -191,23 +192,110 @@ def wrapper(self, *_args, **_kwargs):
         return wrapper
 
     @_check_connected
-    def run_command(self, command):
+    def run(self, command, *, codec="utf-8", decodeerrors="strict",
+            force_tty=False, stderr_merge=False, stderr_loglevel=None,
+            stdout_loglevel=None):
+
         """Run a command over the SSHConnection
 
         Args:
             command (string): The command to run
-
-        Returns:
-            int: exitcode of the command
+            codec (string, optional): output encoding. Defaults to "utf-8".
+            decodeerrors (string, optional): behavior on decode errors. Defaults
+                 to "strict". Refer to stdtypes' bytes.decode for details.
+            force_tty (bool, optional): force allocate a tty (ssh -tt). Defaults
+                 to False
+            stderr_merge (bool, optional): merge ssh subprocess stderr into
+                 stdout. Defaults to False.
+            stdout_loglevel (int, optional): log stdout with specific log level
+                 as well. Defaults to None, i.e. don't log.
+            stderr_loglevel (int, optional): log stderr with specific log level
+                 as well. Defaults to None, i.e. don't log.
+
+        returns:
+            (stdout, stderr, returncode)
         """
+
         complete_cmd = ["ssh"] + self._get_ssh_args()
+        if force_tty:
+            complete_cmd += ["-tt"]
         complete_cmd += [self.host, command]
-        res = subprocess.check_call(
-            complete_cmd,
-            stdin=subprocess.DEVNULL,
-        )
+        self._logger.debug("Sending command: %s", complete_cmd)
+        if stderr_merge:
+            stderr_pipe = subprocess.STDOUT
+        else:
+            stderr_pipe = subprocess.PIPE
+        try:
+            sub = subprocess.Popen(
+                complete_cmd, stdout=subprocess.PIPE, stderr=stderr_pipe,
+                stdin=subprocess.DEVNULL
+            )
+        except:
+            raise ExecutionError(
+                "error executing command: {}".format(complete_cmd)
+            )
 
-        return res
+        stdout = []
+        stderr = []
+
+        readable = {
+            sub.stdout.fileno(): (sub.stdout, stdout, stdout_loglevel),
+        }
+
+        if sub.stderr is not None:
+            readable[sub.stderr.fileno()] = (sub.stderr, stderr, stderr_loglevel)
+
+        while readable:
+            for fd in select(readable, [], [])[0]:
+                stream, output, loglevel = readable[fd]
+                line = stream.readline().decode(codec, decodeerrors)
+                if line == "": # EOF
+                    del readable[fd]
+                else:
+                    line = line.rstrip('\n')
+                    output.append(line)
+                    if loglevel is not None:
+                        self._logger.log(loglevel, line)
+
+        return stdout, stderr, sub.wait()
```

Comment:
Is all of this really necessary? Seems much more compilcated than the SSHDriver.

Reply:
I want the function to be synchronous, (like `subprocess.check_output`/0, but display output in "real-time" (like if stdout/stderr is going to a terminal). If I do it like in the SSHDriver, I won't get output until the process terminates, which is annoying if the process is long-running. 

For that reason, I redirect stdout/stderr into pipes, but now the libc will do block-buffering, thus I provide a way to specify `-tt`, which forces allocation of a PTY, so the output streams are line buffered. As the buffer for stdout/stderr is limited, it must be cleared regularly. A simple `stdout.readline()` followed by a `stderr.readline()` and bail if both are at EOF won't cut it, because it can deadlock: subprocess fills the stderr buffer and blocks while parent waits to read a line from the stdout buffer. For that reason, the `select` was necessary.
`subprocess` uses threads instead behind the scene to achieve this. I guess because it allows a unified implementation for *nix and Windows (Windows' select does only work on sockets and not file descriptors in general).

It might make sense for SSHDriver to do the same as well..